src/helm/benchmark/run_spec.py

```diff
@@ -0,0 +1,89 @@
+from dataclasses import dataclass, field
+import importlib
+import os
+import pkgutil
+from typing import Callable, Dict, Iterable, List, Optional, TypeVar
+
+from helm.benchmark.adaptation.adapter_spec import AdapterSpec
+from helm.benchmark.augmentations.data_augmenter import DataAugmenterSpec
+from helm.benchmark.metrics.metric import MetricSpec
+from helm.benchmark.scenarios.scenario import ScenarioSpec
+
+
+@dataclass(frozen=True)
+class RunSpec:
+    """
+    Specifies how to do a single run, which gets a scenario, adapts it, and
+    computes a list of stats based on the defined metrics.
+    """
+
+    # Unique identifier of the RunSpec
+    name: str
+
+    # Which scenario
+    scenario_spec: ScenarioSpec
+
+    # Specifies how to adapt an instance into a set of requests
+    adapter_spec: AdapterSpec
+
+    # What to evaluate on
+    metric_specs: List[MetricSpec]
+
+    # Data augmenter. The default `DataAugmenterSpec` does nothing.
+    data_augmenter_spec: DataAugmenterSpec = DataAugmenterSpec()
+
+    # Groups that this run spec belongs to (for aggregation)
+    groups: List[str] = field(default_factory=list)
+
+    def __post_init__(self):
+        """
+        `self.name` is used as the name of the output folder for the `RunSpec`.
+        Clean up `self.name` by replacing any "/"'s with "_".
+        """
+        # TODO: Don't mutate name! clean this up before passing it into the constructor here
+        object.__setattr__(self, "name", self.name.replace(os.path.sep, "_"))
+
+
+RunSpecFunction = Callable[..., RunSpec]
+
+
+_REGISTERED_RUN_SPEC_FUNCTIONS: Dict[str, RunSpecFunction] = {}
+"""Dict of run spec function names to run spec functions."""
+
+
+F = TypeVar("F", bound=RunSpecFunction)
+
+
+def run_spec_function(name: str) -> Callable[[F], F]:
```

Comment:
Could we have a general high-level overview of how the registration works?

Reply:
I was wondering the same, so I am going to explain what I understand and @yifanmai correct me if I am wrong.
1. The runner calls `get_run_spec_function` which calls `discover_run_spec_functions`
2. The `discover_run_spec_functions` finds all files in `helm.benchmark.run_specs` and imports them. This defines all get_[NAME}_spec() function thanks to the decorator `@run_spec_function` which stores all functions in `_REGISTERED_RUN_SPEC_FUNCTIONS`.
3. Then, `get_run_spec_function` simply access the element in the dict.

What I am not sure is how the run spec files are not imported several times since `get_run_spec_function` calls every time the discover function without a flag. Does Python handle automatically the fact that it should not import several times the same module?

Reply:
> Does Python handle automatically the fact that it should not import several times the same module?

Yes, the module is only imported the first time. Additionally, importing is thread-safe. Refer to the [Python docs](https://docs.python.org/3/reference/import.html) for more info on how import works.